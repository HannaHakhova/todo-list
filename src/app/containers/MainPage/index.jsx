import React, {Component} from 'react';
import {Grid} from 'semantic-ui-react';
import TodosPart from 'containers/TodosPart';
import TagsPart from 'containers/TagsPart';
import {getAllData, createData, updateData, deleteData, todosTagsApiUrl, todoApiUrl, tagApiUrl} from 'api';
import {getDateNow, getApiDate} from 'utils';
import rand from 'random-key';

class MainPage extends Component {
    constructor() {
        super();
        this.state = {
            todosTags: [],
            todos: [],
            tags: []
        };
    }

    componentDidMount = () => {
        this.getAllData();
    }

    getAllData = () => {
        Promise.all([getAllData(todosTagsApiUrl), getAllData(todoApiUrl), getAllData(tagApiUrl)])
            .then(response => this.setState({todosTags: response[0].data, todos: response[1].data, tags: response[2].data}));
    };

    createTodo = title => {
        if (title.trim()) {
            const newTodo = {id: rand.generate(), title, completed: false, createdDate: getApiDate(getDateNow())};
            const newTodos = [newTodo, ...this.state.todos];
            this.setState({todos: newTodos});
            createData(todoApiUrl, newTodo);
        }
    }

    createTag = title => {
        if (title.trim()) {
            const newTag = {id: rand.generate(), title};
            const newTags = [newTag, ...this.state.tags];
            this.setState({tags: newTags});
            createData(tagApiUrl, newTag);
        }
    }

    deleteTodo = idToBeDeleted => {
        this.setState({todos: this.state.todos.filter(todo => todo.id !== idToBeDeleted)});
        this.setState({todosTags: this.state.todosTags.filter(todoTag => todoTag.taskId !== idToBeDeleted)});
        deleteData(todoApiUrl, idToBeDeleted);
        deleteData(todosTagsApiUrl, idToBeDeleted);
    }

    deleteTag = idToBeDeleted => {
        this.setState({tags: this.state.tags.filter(tag => tag.id !== idToBeDeleted)});
        this.setState({todosTags: this.state.todosTags.filter(todoTag => todoTag.tagId !== idToBeDeleted)});
        deleteData(tagApiUrl, idToBeDeleted);
    };

    deleteTodoTag = idToBeDeleted => {
        this.setState({todosTags: this.state.todosTags.filter(todoTag => todoTag.id !== idToBeDeleted)});
        deleteData(todosTagsApiUrl, idToBeDeleted);
    };

    addTodoTag = (tagId, todoId) => {
        const arrTagId = this.state.todosTags.filter(x => x.tagId === tagId);
        const arrTodoId = arrTagId.filter(x => x.taskId === todoId);
        if (arrTodoId.length === 0) {
            const todoTagIds = this.state.todosTags.map(x => x.id);
            const newTodoTag = {id: Math.max(...todoTagIds) + 1, taskId: todoId, tagId};
            const newTodoTags = [...this.state.todosTags, newTodoTag];
            this.setState({todosTags: newTodoTags});
            createData(todosTagsApiUrl, newTodoTag);
        }
    }

    toggleCompleted = idToBeMarkedAsCompleted => {
        const todoToComplete = this.state.todos.filter(todo => todo.id === idToBeMarkedAsCompleted);
        updateData(todoApiUrl, idToBeMarkedAsCompleted, {completed: !todoToComplete[0].completed})
            .then(response => {
                this.setState({todos: this.state.todos.reduce((previous, current) => {
                    if (current.id === idToBeMarkedAsCompleted) {
                        return [...previous, {...response.data, createdDate: response.data.createdDate}];
                    }
                    return [...previous, current];
                }, [])});
            });
    }

    clearCompleted = () => {
        const {todos, todosTags} = this.state;
        const todoIdsToDeleteArray = todos.filter(todo => todo.completed).map(y => y.id);
        const todoTagIdsToDeleteArray = todosTags.filter(x => todoIdsToDeleteArray.includes(x.taskId)).map(y => y.id);
        const todosPromisesArray = todoIdsToDeleteArray.map(x => deleteData(todoApiUrl, x));


        Promise.all(todosPromisesArray)
            .then(() => {
                this.setState({
                    todos: todos.filter(x => !todoIdsToDeleteArray.includes(x.id)),
                    todosTags: todosTags.filter(x => !todoTagIdsToDeleteArray.includes(x.id))
                });
            })
            .catch(() => {
                this.getAllData();
            });
    }

    render() {
        return (
            <Grid columns='equal'>
                <Grid.Column width={4} />
                <Grid.Column width={6}>
                    <TodosPart
                        todos={this.state.todos}
                        todosTags={this.state.todosTags}
                        tags={this.state.tags}
                        createTodo={this.createTodo}
                        deleteTodo={this.deleteTodo}
                        toggleCompleted={this.toggleCompleted}
                        clearCompleted={this.clearCompleted}
                        deleteTodoTag={this.deleteTodoTag}
                        addTodoTag={this.addTodoTag} />
                </Grid.Column>
                <Grid.Column width={3} >
                    <TagsPart
                        tags={this.state.tags}
                        todosTags={this.state.todosTags}
                        createTag={this.createTag}
                        deleteTag={this.deleteTag} />
                </Grid.Column>
            </Grid>);
    }
}

export default MainPage;
