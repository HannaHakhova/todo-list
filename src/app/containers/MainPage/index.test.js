/* eslint-disable react/jsx-filename-extension */

import React from 'react';
import {shallow} from 'enzyme';
import MainPage from './index';

describe('MainPage', () => {
    it('matches expected snapshot', () => {
        const wrapper = shallow(<MainPage />);
        expect(wrapper).toMatchSnapshot();
    });
});