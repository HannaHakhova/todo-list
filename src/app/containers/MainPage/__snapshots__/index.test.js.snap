// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage matches expected snapshot 1`] = `
<Grid
  columns="equal"
>
  <GridColumn
    width={4}
  />
  <GridColumn
    width={6}
  >
    <TodosPart
      addTodoTag={[Function]}
      clearCompleted={[Function]}
      createTodo={[Function]}
      deleteTodo={[Function]}
      deleteTodoTag={[Function]}
      tags={Array []}
      todos={Array []}
      todosTags={Array []}
      toggleCompleted={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={3}
  >
    <TagsPart
      createTag={[Function]}
      deleteTag={[Function]}
      tags={Array []}
      todosTags={Array []}
    />
  </GridColumn>
</Grid>
`;
